pkg/cli/cli.go

```diff
@@ -0,0 +1,276 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cli
+
+import (
+	"fmt"
+	"log"
+	"os"
+
+	"sigs.k8s.io/kubebuilder/pkg/plugin"
+	"sigs.k8s.io/kubebuilder/pkg/scaffold"
+	"sigs.k8s.io/kubebuilder/pkg/scaffold/project"
+
+	"github.com/spf13/cobra"
+	"github.com/spf13/pflag"
+)
+
+const (
+	noticeColor = "\033[1;36m%s\033[0m"
+)
+
+// Option is a function that can configure the CLI
+type Option func(*CLI) error
+
+// CLI defines the command line structure and interfaces that are used to
+// scaffold kubebuilder project files.
+type CLI struct {
+	commandName    string
+	defProjVersion string
+	cmd            *cobra.Command
+	extraCommands  []*cobra.Command
+	plugins        map[string][]plugin.Base
+}
+
+// New creates a new CLI instance.
+func New(opts ...Option) (*CLI, error) {
+	c := &CLI{
+		commandName:    "kubebuilder",
+		defProjVersion: project.Version2,
+		plugins:        map[string][]plugin.Base{},
+	}
+	for _, opt := range opts {
+		if err := opt(c); err != nil {
+			return nil, err
+		}
+	}
+
+	if err := c.initialize(); err != nil {
+		return nil, err
+	}
+	return c, nil
+}
+
+// Run runs the CLI.
+func (c CLI) Run() error {
+	return c.cmd.Execute()
+}
+
+// WithCommandName is an Option that sets the CLI's root command name.
+func WithCommandName(name string) Option {
+	return func(c *CLI) error {
+		c.commandName = name
+		return nil
+	}
+}
+
+// WithDefaultProjectVersion is an Option that sets the CLI's default project
+// version. Setting an unknown version will result in an error.
+func WithDefaultProjectVersion(version string) Option {
+	return func(c *CLI) error {
+		c.defProjVersion = version
+		return nil
+	}
+}
+
+// WithPlugins is an Option that sets the CLI's plugins.
+func WithPlugins(plugins ...plugin.Base) Option {
+	return func(c *CLI) error {
+		for _, p := range plugins {
+			ver := p.Version()
+			if _, ok := c.plugins[ver]; !ok {
+				c.plugins[ver] = []plugin.Base{}
+			}
+			c.plugins[ver] = append(c.plugins[ver], p)
+		}
+		return nil
+	}
+}
+
+// WithExtraCommands is an Option that adds extra subcommands to the CLI.
+// Adding extra commands that duplicate existing commands results in an error.
+func WithExtraCommands(cmds ...*cobra.Command) Option {
+	return func(c *CLI) error {
+		c.extraCommands = cmds
+		return nil
+	}
+}
+
+// initialize initializes the CLI.
+func (c *CLI) initialize() error {
+	// Inject the command name into each plugin that supports
+	// injectable command names.
+	for _, ps := range c.plugins {
+		for _, p := range ps {
+			if in, ok := p.(plugin.InjectsCommandName); ok {
+				in.InjectCommandName(c.commandName)
+			}
+		}
+	}
+
+	// Check the current working directory for a PROJECT file.
+	// Return whether it was found, and if so, what project version
+	// it is for.
+	foundProject, projectVersion := getProjectVersion()
```

Comment:
seems like we should be loading this once and keeping it around (and potentially passing to plugins, because we don't want plugins to have to implement this individually, which'd be a mess for updating)

Reply:
I assume you mean passing a `ProjectSpec` (or similarly named) struct that is read in from the `PROJECT` file?

Reply:
I'm working on centrallizing the configuration file (#1325). I was using `internal/config` as the package to prevent third party to modify the file, but as I see you do need this config, I will probably expose the struct. IMHO, `pkg/model/config` seems to be a good point to have a `Config` struct with the information from that file.

Reply:
I will merge #1325 here one it gets some reviews.